// @saber2pr/typedoc-plugin-localization-zh/dist/components/convert-component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConvertComponent = void 0;
const typedoc_1 = require("typedoc");
const file_operations_1 = require("../utils/file-operations");
const class_factory_1 = require("../utils/factories/class-factory");
const enum_factory_1 = require("../utils/factories/enum-factory");
const parser_1 = require("../utils/parser");
const constants_1 = require("../utils/constants");
const interface_factory_1 = require("../utils/factories/interface-factory");
const function_factory_1 = require("../utils/factories/function-factory");
class ConvertComponent {
    constructor(application) {
        /**
         * Global functions data.
         */
        this.globalFuncsJson = {};
        this.application = application;
        this.initialize();
    }
    initialize() {
        this.application.converter.on(typedoc_1.Converter.EVENT_RESOLVE, this.resolve.bind(this));
        this.application.converter.on(typedoc_1.Converter.EVENT_RESOLVE_BEGIN, this.onResolveBegin.bind(this));
        this.application.converter.on(typedoc_1.Converter.EVENT_RESOLVE_END, this.onResolveEnd.bind(this));
        this.application.converter.on(typedoc_1.Converter.EVENT_END, this.onEnd.bind(this));
        this.application.converter.on(typedoc_1.Converter.EVENT_BEGIN, this.onBegin.bind(this));
        this.parser = new parser_1.Parser();
        this.fileOperations = new file_operations_1.FileOperations(this.application.logger);
    }
    /**
     * Executes when the file convertion begins.
     */
    onBegin() {
        /**
         * Command line arguments
         */
        const options = this.application.options.getRawValues();
        /**
         * Get json's directory. Where json files should be exported.
         */
        this.mainDirToExport = options[constants_1.Constants.CONVERT_OPTION];
        if (!this.fileOperations.ifDirectoryExists(this.mainDirToExport)) {
            this.fileOperations.createDir(this.mainDirToExport);
        }
    }
    /**
     * Execute when the convertion of the files end.
     */
    onEnd() {
        /**
         * Stop the process after all json's are created.
         *
         * It isn't necessary to continue execution because json's have to be translated first
         * then execute the generation of the documentation.
         */
        process.exit(0);
    }
    onResolveBegin(context) {
        /**
         * Create main file which would contains all global functions.
         */
        this.fileOperations.createFile(this.mainDirToExport, null, constants_1.Constants.GLOBAL_FUNCS_FILE_NAME, 'json');
        this.fileOperations.appendFileData(this.mainDirToExport, null, constants_1.Constants.GLOBAL_FUNCS_FILE_NAME, 'json', {});
    }
    onResolveEnd() {
        /**
         * Write the last built json file.
         *
         * It happens here because we are unable to track into the
         * @reslove handler when exactly the execution stops.
         * And we know for sure that the last step before stropping the conversion is here.
         */
        if (this.factoryInstance && !this.factoryInstance.isEmpty()) {
            const filePath = this.reflection.sources[0].fileName;
            const processedDir = this.fileOperations.getFileDir(filePath);
            this.fileOperations.createFile(this.mainDirToExport, processedDir, this.jsonObjectName, 'json');
            this.fileOperations.appendFileData(this.mainDirToExport, processedDir, this.jsonObjectName, 'json', this.factoryInstance.getJsonContent());
        }
        /**
         * Write all collected data for all global functions into corresponding file.
         */
        const funcObjKeys = Object.keys(this.globalFuncsJson);
        if (funcObjKeys.length) {
            this.fileOperations.appendFileData(this.mainDirToExport, null, constants_1.Constants.GLOBAL_FUNCS_FILE_NAME, 'json', this.globalFuncsJson);
        }
    }
    /**
     * Triggers per every reflection object.
     * @param context
     * @param reflection
     */
    resolve(context, reflection) {
        switch (reflection.kind) {
            case typedoc_1.ReflectionKind.Enum:
            case typedoc_1.ReflectionKind.Class:
            case typedoc_1.ReflectionKind.Interface:
                /**
                 * Writes file content when the resolve process for to Object ends
                 * per(Class, Enum, Interface).
                 */
                if (this.jsonObjectName !== reflection.name && this.jsonObjectName !== undefined) {
                    if (!this.factoryInstance.isEmpty()) {
                        const filePath = this.reflection.sources[0].fileName;
                        const processedDir = this.fileOperations.getFileDir(filePath);
                        this.fileOperations.createFile(this.mainDirToExport, processedDir, this.jsonObjectName, 'json');
                        this.fileOperations.appendFileData(this.mainDirToExport, processedDir, this.jsonObjectName, 'json', this.factoryInstance.getJsonContent());
                    }
                }
                const data = this.getCommentInfo(reflection);
                this.jsonObjectName = reflection.name;
                this.reflection = reflection;
                this.factoryInstance = this.instanceBuilder(reflection.kind, reflection.name);
                this.factoryInstance.buildObjectStructure(data);
                break;
            case typedoc_1.ReflectionKind.Property:
            case typedoc_1.ReflectionKind.CallSignature:
            case typedoc_1.ReflectionKind.EnumMember:
                /**
                 * Skip reflections with type @ReflectionKind.Function because they are aslo @ReflectionKInd.CallSignature
                 * but the handling process here is not appropriate for them.
                 */
                if (reflection.parent.kind === typedoc_1.ReflectionKind.Function) {
                    break;
                }
                const getData = this.getCommentInfo(reflection);
                this.factoryInstance.appendAttribute(this.jsonObjectName, reflection.kind, reflection.name, getData);
                break;
            case typedoc_1.ReflectionKind.Function:
                const funcData = this.getCommentInfo(reflection.signatures[0]);
                const funcInstance = this.instanceBuilder(reflection.kind, reflection.name);
                funcInstance.buildObjectStructure(funcData);
                if (!funcInstance.isEmpty()) {
                    this.globalFuncsJson = Object.assign(funcInstance.getJsonContent(), this.globalFuncsJson);
                }
                break;
            case typedoc_1.ReflectionKind.GetSignature:
            case typedoc_1.ReflectionKind.SetSignature:
                const accessorName = reflection.parent.name;
                const accessorType = reflection.kind;
                const accessorData = this.getCommentInfo(reflection);
                this.factoryInstance.appendAccessorAttributes(this.jsonObjectName, reflection.parent.kind, accessorName, accessorType, accessorData);
            default:
                return;
        }
    }
    /**
     * Returns all comment info including tags and parameters.
     * @param reflection
     */
    getCommentInfo(reflection) {
        const options = this.application.options.getRawValues();
        if (!reflection.comment) {
            return;
        }
        let comment = this.getCommentData(reflection.comment);
        if (options[constants_1.Constants.INCLUDE_TAGS_OPTION] && reflection.comment.tags) {
            comment[constants_1.Constants.COMMENT] = Object.assign(this.getTagsComments(reflection.comment), comment[constants_1.Constants.COMMENT]);
        }
        if (options[constants_1.Constants.INCLUDE_PARAMS_OPTION] && reflection.parameters) {
            comment[constants_1.Constants.COMMENT] = Object.assign(this.getParamsComments(reflection), comment[constants_1.Constants.COMMENT]);
        }
        if (options[constants_1.Constants.INCLUDE_TAGS_OPTION] && reflection.comment.returns) {
            comment[constants_1.Constants.COMMENT] = Object.assign(this.getReturnsComment(reflection), comment[constants_1.Constants.COMMENT]);
        }
        return comment;
    }
    /**
     * Extract the comment from the @return parameter.
     * @param reflection
     */
    getReturnsComment(reflection) {
        let returns = {};
        returns[constants_1.Constants.RETURNS] = this.getCommentData(reflection.comment);
        return returns;
    }
    /**
     * Returns all parameters per comment.
     * @param reflection
     */
    getParamsComments(reflection) {
        let params = {};
        params[constants_1.Constants.PARAMS] = {};
        reflection.parameters.forEach(param => {
            if (!param.comment) {
                return;
            }
            const paramComment = this.getCommentData(param.comment);
            const paramCommentKeys = Object.keys(paramComment[constants_1.Constants.COMMENT]);
            if (paramCommentKeys.length) {
                params[constants_1.Constants.PARAMS][param.name] = paramComment;
            }
        });
        return Object.keys(params[constants_1.Constants.PARAMS]).length ? params : {};
    }
    /**
     * Returns all tags per comment.
     * @param comment
     */
    getTagsComments(comment) {
        let tags = {};
        tags[constants_1.Constants.TAGS] = {};
        comment.tags.forEach(tag => {
            let tagComment = this.getCommentData(tag);
            if (tag.tagName) {
                tags[constants_1.Constants.TAGS][tag.tagName] = tagComment;
            }
        });
        return tags;
    }
    /**
     * Returns comment content.
     * @param obj
     */
    getCommentData(obj) {
        const comment = {};
        comment[constants_1.Constants.COMMENT] = {};
        let splittedObj;
        if (obj.text && obj.text.trim().length) {
            splittedObj = this.parser.splitByCharacter(obj.text, '\n');
            comment[constants_1.Constants.COMMENT][constants_1.Constants.TEXT] = splittedObj;
        }
        if (obj.shortText && obj.shortText.trim().length) {
            splittedObj = this.parser.splitByCharacter(obj.shortText, '\n');
            comment[constants_1.Constants.COMMENT][constants_1.Constants.SHORT_TEXT] = splittedObj;
        }
        if (obj.tagName) {
            comment[constants_1.Constants.COMMENT][constants_1.Constants.TAG_NAME] = obj.tagName;
        }
        if (obj.returns) {
            comment[constants_1.Constants.COMMENT][constants_1.Constants.RETURNS] = obj.returns;
        }
        return comment;
    }
    /**
     * Builds a instance depending on the Object type.
     * @param objectType
     * @param objectName
     */
    instanceBuilder(objectType, objectName) {
        switch (objectType) {
            case typedoc_1.ReflectionKind.Enum:
                return new enum_factory_1.EnumFactory(objectName);
            case typedoc_1.ReflectionKind.Interface:
                return new interface_factory_1.InterfaceFactory(objectName);
            case typedoc_1.ReflectionKind.Function:
                return new function_factory_1.FunctionFactory(objectName);
            case typedoc_1.ReflectionKind.Class:
                return new class_factory_1.ClassFactory(objectName);
            default:
                null;
        }
    }
}
exports.ConvertComponent = ConvertComponent;
