// @saber2pr/typedoc-plugin-localization-zh/dist/components/render-component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderComponenet = void 0;
const path = require("path");
const typedoc_1 = require("typedoc");
const file_operations_1 = require("../utils/file-operations");
const json_keys_1 = require("../utils/enums/json-keys");
const constants_1 = require("../utils/constants");
const parser_1 = require("../utils/parser");
class RenderComponenet {
    constructor(application) {
        this.application = application;
        this.initialize();
    }
    initialize() {
        this.application.renderer.on(typedoc_1.RendererEvent.BEGIN, this.onRenderBegin.bind(this));
        this.fileOperations = new file_operations_1.FileOperations(this.application.logger);
        this.parser = new parser_1.Parser();
    }
    onRenderBegin(event) {
        event.project.localization = this.application.options.getValue("localize");
        const reflections = event.project.reflections;
        const options = this.application.options.getRawValues();
        const localizeOpt = options[constants_1.Constants.RENDER_OPTION];
        this.warns = options[constants_1.Constants.WARNS];
        if (localizeOpt) {
            this.mainDirOfJsons = localizeOpt;
            this.globalFuncsData = this.fileOperations.getFileData(this.mainDirOfJsons, constants_1.Constants.GLOBAL_FUNCS_FILE_NAME, 'json');
            this.runCommentReplacements(reflections);
        }
    }
    runCommentReplacements(reflections) {
        const keys = Object.keys(reflections);
        keys.forEach(key => {
            const reflection = reflections[key];
            this.processTheReflection(reflection);
        });
    }
    processTheReflection(reflection) {
        switch (reflection.kind) {
            case typedoc_1.ReflectionKind.Class:
            case typedoc_1.ReflectionKind.Enum:
            case typedoc_1.ReflectionKind.Interface:
                const filePath = reflection.sources[0].fileName;
                let processedDir = this.mainDirOfJsons;
                const parsedPath = this.fileOperations.getFileDir(filePath);
                if (parsedPath) {
                    processedDir = path.join(processedDir, parsedPath);
                }
                this.data = this.fileOperations.getFileData(processedDir, reflection.name, 'json');
                if (this.data) {
                    this.updateComment(reflection, this.data[reflection.name]);
                }
                break;
            case typedoc_1.ReflectionKind.Property:
            case typedoc_1.ReflectionKind.CallSignature:
            case typedoc_1.ReflectionKind.EnumMember:
                /**
                  * Skip reflections with type @ReflectionKind.Function because they are aslo @ReflectionKInd.CallSignature
                  * but the handling process here is not appropriate for them.
                  */
                if (reflection.parent === typedoc_1.ReflectionKind.Function) {
                    break;
                }
                const parent = this.getParentBasedOnType(reflection, reflection.kind);
                const parentName = parent.name;
                const attributeName = reflection.name;
                const attributeData = this.getAttributeData(parentName, json_keys_1.AttributeType[reflection.kind], attributeName);
                if (attributeData) {
                    this.updateComment(reflection, attributeData);
                }
                break;
            case typedoc_1.ReflectionKind.Function:
                if (!this.globalFuncsData) {
                    break;
                }
                const funcName = reflection.name;
                const funcData = this.globalFuncsData[funcName];
                this.updateComment(reflection.signatures[0], funcData);
                break;
            case typedoc_1.ReflectionKind.GetSignature:
            case typedoc_1.ReflectionKind.SetSignature:
                const accessorParent = this.getParentBasedOnType(reflection, reflection.kind);
                const accessor = reflection.parent;
                const accessorSignature = reflection.kind;
                const data = this.getAccessorAttributeData(accessorParent.name, json_keys_1.AttributeType[accessor.kind], accessor.name, json_keys_1.AttributeType[accessorSignature]);
                if (data) {
                    this.updateComment(reflection, data);
                }
                break;
            default:
                return;
        }
    }
    getAttribute(parentName, attribute) {
        if (this.data && this.data[parentName]) {
            return this.data[parentName][attribute];
        }
    }
    getAttributeData(parentName, attribute, attributeName) {
        const data = this.getAttribute(parentName, attribute);
        if (data) {
            return data[attributeName];
        }
    }
    getAccessorAttributeData(parentName, attribute, attributeName, accessorType) {
        const data = this.getAttributeData(parentName, attribute, attributeName);
        if (data) {
            return data[accessorType];
        }
    }
    updateComment(reflection, dataObj) {
        if (!reflection.comment || (dataObj && !dataObj[constants_1.Constants.COMMENT])) {
            return;
        }
        let parsed;
        if (reflection.comment.text) {
            parsed = this.parser.joinByCharacter(dataObj[constants_1.Constants.COMMENT][constants_1.Constants.TEXT], '\n');
            reflection.comment.text = parsed;
        }
        if (reflection.comment.shortText) {
            parsed = this.parser.joinByCharacter(dataObj[constants_1.Constants.COMMENT][constants_1.Constants.SHORT_TEXT], '\n');
            reflection.comment.shortText = parsed;
        }
        if (reflection.comment.returns) {
            parsed = this.parser.joinByCharacter(dataObj[constants_1.Constants.COMMENT][constants_1.Constants.RETURNS], '\n');
            reflection.comment.returns = parsed;
        }
        if (reflection.comment.tags && dataObj[constants_1.Constants.COMMENT][constants_1.Constants.TAGS]) {
            reflection.comment.tags.forEach(tag => {
                const tagFromJson = dataObj[constants_1.Constants.COMMENT][constants_1.Constants.TAGS][tag.tagName];
                try {
                    tag.tagName = tagFromJson[constants_1.Constants.COMMENT].tagName;
                    tag.text = this.parser.joinByCharacter(tagFromJson[constants_1.Constants.COMMENT].text, '\n');
                }
                catch (e) {
                    if (this.warns) {
                        this.application.logger.log(`Could not find ${tag.tagName} tag of ${reflection.parent.name} in ${reflection.parent.parent.name}`, typedoc_1.LogLevel.Warn);
                    }
                }
            });
        }
        if (reflection.parameters && dataObj[constants_1.Constants.COMMENT][constants_1.Constants.PARAMS]) {
            reflection.parameters.forEach(param => {
                const paramFromJson = dataObj[constants_1.Constants.COMMENT][constants_1.Constants.PARAMS][param.name];
                try {
                    param.comment.text = this.parser.joinByCharacter(paramFromJson[constants_1.Constants.COMMENT].text, '\n');
                }
                catch (e) {
                    if (this.warns) {
                        this.application.logger.log(`Could not find ${param.name} parameter of ${reflection.parent.name} in ${reflection.parent.parent.name}`, typedoc_1.LogLevel.Warn);
                    }
                }
            });
        }
    }
    /**
     * Returns the parent(Class, Interface, Enum) per every Method, Property, Getter, Setter etc.
     * @param reflection
     * @param kind
     */
    getParentBasedOnType(reflection, kind) {
        if (kind === typedoc_1.ReflectionKind.CallSignature ||
            kind === typedoc_1.ReflectionKind.GetSignature ||
            kind === typedoc_1.ReflectionKind.SetSignature) {
            return reflection.parent.parent;
        }
        return reflection.parent;
    }
}
exports.RenderComponenet = RenderComponenet;
