// @saber2pr/typedoc-plugin-localization-zh/dist/components/theme-component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThemeComponent = void 0;
const typedoc_1 = require("typedoc");
const global_funcs_1 = require("../utils/global-funcs");
const template_strings_1 = require("../utils/template-strings");
class ThemeComponent {
    constructor(app) {
        this.app = app;
        this.initialize();
    }
    initialize() {
        this.app.renderer.on(typedoc_1.RendererEvent.BEGIN, this.onRenderBegin.bind(this));
    }
    onRenderBegin(event) {
        if (!event.project.groups) {
            return;
        }
        this.localizeGroupTitles(event.project.groups);
        this.run(event.project.reflections);
    }
    run(reflections) {
        const keys = Object.keys(reflections);
        keys.forEach(key => {
            const reflection = reflections[key];
            this.localizeReflectionDeffinitions(reflection);
        });
    }
    /**
     * Triggers localizaiton of template groups(Accessors, Constructor, Methods, Properties, etc.)
     * and Object abbreviations(Class, Interface, Enum).
     * @param reflection
     */
    localizeReflectionDeffinitions(reflection) {
        if (reflection.kind === typedoc_1.ReflectionKind.Class ||
            reflection.kind === typedoc_1.ReflectionKind.Enum ||
            reflection.kind === typedoc_1.ReflectionKind.Interface) {
            if (reflection.groups) {
                this.localizeGroupTitles(reflection.groups);
            }
            this.localizeReflectionAbbriviation(reflection);
        }
        if (reflection.parameters && reflection.parameters.length) {
            reflection.parameters.forEach(e => {
                this.localizeParameterFlags(e);
            });
        }
    }
    /**
     * Localize Object abbreviatons(Class, Interface, Enum)
     * @param reflection
     */
    localizeReflectionAbbriviation(reflection) {
        reflection.kindString = this.getLocaleValue(reflection.kindString);
    }
    /**
     * Localize template groups like (Accessors, Methods, Properties, etc.)
     * @param groups
     */
    localizeGroupTitles(groups) {
        groups.forEach(element => {
            element.title = this.getLocaleValue(element.title);
        });
    }
    localizeParameterFlags(param) {
        if (param.flags && param.flags.length) {
            param.flags.forEach((f, idx) => {
                param.flags[idx] = this.getLocaleValue(f);
            });
        }
    }
    /**
     * Get template localized string from cached translations.
     * @param value
     */
    getLocaleValue(value) {
        return global_funcs_1.GlobalFuncs.getKeyValuePairVal(template_strings_1.HardcodedStrings.getTemplateStrings(), template_strings_1.HardcodedStrings.getLocal(), value);
    }
}
exports.ThemeComponent = ThemeComponent;
