// @saber2pr/typedoc-plugin-localization-zh/dist/utils/file-operations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileOperations = void 0;
const fs = require("fs-extra");
const path = require("path");
class FileOperations {
    constructor(logger) {
        this.logger = logger;
    }
    createDir(dirName) {
        try {
            fs.mkdirpSync(dirName);
        }
        catch (err) {
            this.logger.error(err);
            return false;
        }
        return true;
    }
    ifDirectoryExists(dirName) {
        if (fs.existsSync(dirName)) {
            if (!fs.statSync(dirName).isDirectory()) {
                this.logger.error(`The output target exists ${dirName} but it is not a directory!`);
                return false;
            }
            return true;
        }
        return false;
    }
    ifFileExists(filePath) {
        if (fs.existsSync(filePath)) {
            if (!fs.statSync(filePath).isFile()) {
                this.logger.error(`The ouput targets exists ${path} but it is not a file!`);
                return false;
            }
            return true;
        }
        return false;
    }
    removeDirectoryOrFile(path) {
        try {
            fs.removeSync(path);
        }
        catch (err) {
            this.logger.warn(`Could not clear the directory or path: ${path}!`);
            return false;
        }
        return true;
    }
    openFileSync(dirPath, fileName) {
        if (this.ifDirectoryExists(dirPath)) {
            const filePath = path.join(dirPath, fileName);
            return fs.openSync(filePath, 'w+');
        }
        this.logger.error(`Direcotry with path: ${dirPath} does not exists!`);
        return null;
    }
    closeFileSync(fileReference) {
        if (!!fileReference) {
            fs.closeSync(fileReference);
        }
        else {
            this.logger.error(`Invalid faile reference: ${fileReference}`);
        }
    }
    prepareOutputDirectory(mainDir, fileObjects) {
        fileObjects.forEach(element => {
            const dirToExport = this.constructFilePath(mainDir, element.fileName);
            if (!dirToExport) {
                return;
            }
            if (!this.ifDirectoryExists(dirToExport)) {
                this.createDir(dirToExport);
            }
        });
    }
    getFileDir(filePath) {
        const parsedPath = path.parse(filePath);
        return parsedPath.dir;
    }
    appendFileData(mainDir, filePath, fileName, extension, data) {
        let path = this.constructFilePath(mainDir, filePath);
        const currentFileFd = this.openFileSync(path, `${fileName}.${extension}`);
        fs.writeSync(currentFileFd, JSON.stringify(data, null, 4), null, 'utf8');
        this.closeFileSync(currentFileFd);
    }
    createFile(mainDir, filePath, fileName, fileExtension) {
        let constructedFilePath = this.constructFilePath(mainDir, filePath);
        const file = path.join(constructedFilePath, `${fileName}.${fileExtension}`);
        if (!this.ifFileExists(file)) {
            fs.createFileSync(file);
        }
    }
    getFileData(filePath, fileName, fileExtension) {
        const jsonFilePath = path.join(filePath, `${fileName}.${fileExtension}`);
        if (!this.ifDirectoryExists(filePath) || !this.ifFileExists(jsonFilePath)) {
            return null;
        }
        return fs.readJsonSync(jsonFilePath);
    }
    constructFilePath(mainDir, filePath) {
        const processedPath = filePath ? path.normalize(filePath).toString() : filePath;
        let filePathBuilder = mainDir;
        if (processedPath) {
            filePathBuilder = path.join(filePathBuilder, processedPath);
        }
        return filePathBuilder;
    }
}
exports.FileOperations = FileOperations;
